File: 5. Prefix Sum/Forest_Queries.py
n, q = map(int, input().split())
forest = [input().strip() for _ in range(n)]
queries = [tuple(map(int, input().split())) for _ in range(q)]

prefix = [[0] * (n + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    for j in range(1, n + 1):
        current_cell_tree = 1 if forest[i - 1][j - 1] == '*' else 0
        top_prefix = prefix[i - 1][j]
        left_prefix = prefix[i][j - 1] 
        overlap_prefix = prefix[i - 1][j - 1]

        prefix[i][j] = (current_cell_tree + top_prefix + left_prefix - overlap_prefix)

for query in queries:
    y1, x1, y2, x2 = query

    total_area = prefix[y2][x2]
    top_strip = prefix[y1 - 1][x2]
    left_strip = prefix[y2][x1 - 1]
    overlap_corner = prefix[y1 - 1][x1 - 1]

    trees = (total_area - top_strip - left_strip + overlap_corner)
    print(trees)
    

File: Prefix Sum/Forest_Queries.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int main()
{
    int n, q;
    cin >> n >> q;

    
    vector<string> forest(n);
    for (int i = 0; i < n; ++i)
    {
        cin >> forest[i];
    }


    vector<vector<int>> prefix(n + 1, vector<int>(n + 1, 0));

    for (int row = 1; row <= n; ++row)
    {
        for (int col = 1; col <= n; ++col)
        {
            int current_cell_tree = (forest[row - 1][col - 1] == '*') ? 1 : 0;
            int top_prefix = prefix[row - 1][col];
            int left_prefix = prefix[row][col - 1];
            int overlap_prefix = prefix[row - 1][col - 1];

            prefix[row][col] = current_cell_tree + top_prefix + left_prefix - overlap_prefix;
        }
    }

    for (int i = 0; i < q; ++i)
    {
        int y1, x1, y2, x2;
        cin >> y1 >> x1 >> y2 >> x2;

        int total_area = prefix[y2][x2];
        int top_strip = prefix[y1 - 1][x2];
        int left_strip = prefix[y2][x1 - 1];
        int overlap_corner = prefix[y1 - 1][x1 - 1];

        int trees_in_rectangle = total_area - top_strip - left_strip + overlap_corner;

        cout << trees_in_rectangle << "\n";
    }

    return 0;
}
